port/osx/omrintrospect.c

```diff
@@ -300,9 +321,75 @@ getThreadContext(J9ThreadWalkState *state)
 
 	if (0 == ret) {
 		/* Wait for the signal handler to complete. */
-		if (0 == read(pipeFileDescriptor[0], buffer, 1)) {
+		ret = timedWait(timeout(state->deadline1));
+		if (TIMEOUT == ret) {
+			RECORD_ERROR(state, TIMEOUT, 0);
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * Utility function to calculate remaining time before the deadline elapses.
+ *
+ * @param deadline system time by which processing should have completed
+ *
+ * @return number of seconds before timeout.
+ */
+static int32_t
+timeout(int64_t deadline)
+{
+	int32_t secs = 0;
+	struct timespec spec;
+	if (-1 == clock_gettime(CLOCK_REALTIME, &spec)) {
+		return 0;
+	}
+
+	secs = (int32_t)(deadline - spec.tv_sec);
+	if (secs > 0) {
+		return secs;
+	}
+
+	return 0;
+}
+
+/**
+ * This function waits for data on a pipe and will block for at most `seconds` seconds.
+ *
+ * @param seconds the maximum number of seconds to block for, must be positive
+ *
+ * @return 0 if success, -1 if error, TIMEOUT if timeout occurred.
+ */
+static int32_t
+timedWait(int32_t seconds)
+{
+	char buffer[1];
+	struct pollfd fds[1];
```

Comment:
Why arrays of length 1? Why not just plain variables?
Please also move declarations into the smallest block:
```
	int32_t ret = 0;

	if (seconds > 0) {
		struct pollfd fds;

		fds.fd = pipeFileDescriptor[0];
		fds.events = POLLIN;

		ret = poll(&fds, 1, seconds * 1000);

		if ((-1 == ret) || (0 != (fds.revents & (POLLHUP | POLLERR | POLLNVAL)))) {
			/* error from poll() */
			ret = -1;
		} else if (0 == ret) {
			/* timeout */
			ret = TIMEOUT;
		} else {
			char buffer;

			if (1 == read(fds.fd, &buffer, 1)) {
				/* success */
				ret = 0;
			} else {
				ret = -1;
			}
		}
	} else {
		/* already timed out */
		ret = TIMEOUT;
	}

	return ret;
```